### ISSUE TITLE
[4.0] StacticBody3D not detected into an Area

### ISSUE THREAD

**Issue by kevinloustau** [godotengine/godot#57539]
### Godot version

v4.0.alpha.custom_build [2aee84c75]

### System information

Windows 11

### Issue description

Moving a StaticBody3D into an Area. I can't detect the body into the Area
It is working correctly using 3.4.2, but not on master at this date.

![should_print](https://user-images.githubusercontent.com/4235617/152127677-868ae387-6972-4e17-a9cc-13c6e1f94535.gif)

### Steps to reproduce

- Move a StaticBody3D programmatically into an area
- print a message from the area signal "on_body_entered"

### Minimal reproduction project

[StaticBody and Area 4.0.zip](https://github.com/godotengine/godot/files/7985044/StaticBody.and.Area.4.0.zip)
[StaticBody and Area 3.4.2.zip](https://github.com/godotengine/godot/files/7985046/StaticBody.and.Area.3.4.2.zip)

**Comment by lawnjelly** [comment]
This is highly likely because the shared BVH has got behind in 4.0, and is pending an update to get them back into sync, there are some bug fixes in 3.x that aren't available in 4.0 yet. I'll be doing this soon.

There's also a longstanding bug with areas and statics (that pre-dates the BVH) that I will hopefully be fixing soon.

**Comment by lawnjelly** [comment]
I've just tried this with #57630 and it doesn't show the message even with the updated BVH. I'll try and have a debug but I'm not sure of what other changes there may be between the two version in the physics.

**Comment by rburing** [comment]
In the minimal reproduction project, the `position.y -= delta * 2.0` code on the box can be moved from `_process` to `_physics_process`. With this change, still neither area detects the static body entering. However, when the constant `2.0` is replaced by something in the range between `41` and `128`, only the *first* area detects the static body entering. When the constant `2.0` is replaced by something in the range between `164` and `251`, only the *second* area detects the static body entering. I hope this can help to figure out what's going on here.

**Comment by rburing** [comment]
The current design of (the code of) Godot Physics is that unmoved areas do not collide with static bodies. In the minimal reproduction project, the internal `GodotAreaPair3D` object (pairing the area and the body) gets created and hence added as a constraint to the body and to the area, but the Godot Physics step `GodotStep3D::step` just ignores it: it only considers the constraints of *moved* areas and *active* bodies (and static bodies are never active), so no collision is detected. The reason the higher velocity caused the collision to be detected is that the areas are considered "moved" on the very first frame (and the velocity was such that the collision happened on the first frame).

In 3.x with Godot Physics the behavior is the same; no collision is detected after the first frame (though the `AreaPairSW` object gets created only after changing the `tree_collision_mask` to `3` in the `BroadPhaseSW`). In 3.x with Bullet Physics the collisions *are* detected (by design of the Bullet Physics server), which explains the discrepancy in the original report.

To address this issue either the design of Godot Physics should be changed (to somehow consider more areas for collision, with the restriction that it should still be efficient), or this caveat should be added to the documentation of every relevant method and signal.

**Comment by lawnjelly** [comment]
Ah :grinning:
I'm so dumb I think the reason this isn't working is by design.

The object you are moving is a `static body`. In physics the word static is used to refer to objects which _should not move_. If you are moving it, it is not meant to detect collisions. If you want to move an object and detect collisions, then it should be a kinematic body, a rigid body, or an area.

The current behaviour therefore makes total sense, and it maybe does do a collision detection on the first entry into the world.

The fact it works in bullet is likely because bullet has no distinction between static and dynamic objects (in a sense it works by accident, but it is not intended).

**Comment by lawnjelly** [comment]
I've marked this as a good first issue : modifying the docs to make clear that statics are intended to be non-moving, and are not guaranteed to detect collisions when moved.

**Comment by mohamedAdhamc** [comment]
> I've marked this as a good first issue : modifying the docs to make clear that statics are intended to be non-moving, and are not guaranteed to detect collisions when moved.

Yes, I actually would like to do just that. Do I have to be assigned or can I just make a pull request?

**Comment by lawnjelly** [comment]
> Yes, I actually would like to do just that. Do I have to be assigned or can I just make a pull request?

No need to be assigned, just make a PR. :+1:

If it was a large PR sometimes it is worth making a note that you are working on it so multiple people don't duplicate work at the same time, but for small ones like this there is usually no need.

**Comment by mohamedAdhamc** [comment]
> > Yes, I actually would like to do just that. Do I have to be assigned or can I just make a pull request?
> 
> No need to be assigned, just make a PR. 👍
> 
> If it was a large PR sometimes it is worth making a note that you are working on it so multiple people don't duplicate work at the same time, but for small ones like this there is usually no need.

Hey excuse my ignorance but would you have an idea as to where in the documentation the description of staticbody3d is, because I cant seem to find it.

**Comment by mohamedAdhamc** [comment]
> The file you need to modify is [`doc/classes/StaticBody3D.xml` in the `master` branch](https://github.com/godotengine/godot/blob/master/doc/classes/StaticBody3D.xml).
> 
> See [Contributing to the class reference](https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html) for more information.

thanks will work on it

**Comment by mohamedAdhamc** [comment]
done

**Comment by AttackButton** [comment]
@kevinloustau as I did comment in https://github.com/godotengine/godot/pull/58710#issuecomment-1136097515, in godot 4.0, you need to use the AnimatableBody instead of the StaticBody when your intention is to move the object.

**Comment by tinmanjuggernaut** [comment]
> I think the reason this isn't working is by design.

This is a bug, not a documentation issue. The behavior is different from Godot 3. I have static bodies that do not move. I have areas attached to kinematic bodies, and the areas do not detect when the static bodies enter them. It worked fine in Godot 3. Imagine a switch on a wall, and the player detects it has an interactable object in front of it. Currently only my kinematic bodies are detected. All of my pickup items (weapons, shields, potions) and interactable nodes are broken.

This issue isn't about StaticBody and whether it moves or not, it's about Area3D. The whole purpose of Area3Ds is to detect physics bodies without calculating collision. Area detection should work with all physics bodies if the Area is monitoring, the collisionshape is enabled, and the collision layer and layer mask line up.

Area3D Manual page - Nothing about static/kinematic/etc as that is irrelevant.
```
3D area that detects CollisionObject3D nodes overlapping, entering, or exiting. 
```
